clippy_lints/src/loops.rs

```diff
@@ -2241,3 +2265,67 @@ impl<'a, 'tcx> Visitor<'tcx> for VarCollectorVisitor<'a, 'tcx> {
         NestedVisitorMap::None
     }
 }
+
+fn check_needless_collect<'a, 'tcx>(expr: &'tcx Expr, cx: &LateContext<'a, 'tcx>) {
+    if let ExprKind::MethodCall(ref method, _, ref args) = expr.node {
+        if let ExprKind::MethodCall(ref chain_method, _, _) = args[0].node {
+            if chain_method.ident.name == "collect" && match_trait_method(cx, &args[0], &paths::ITERATOR) {
+                if method.ident.name == "len" {
+                    span_lint_and_sugg(
+                        cx,
+                        NEEDLESS_COLLECT,
+                        expr.span,
+                        "you are collecting an iterator to check its length",
+                        "consider replacing with",
+                        generate_needless_collect_len_sugg(&args[0], cx),
+                    );
+                }
+                if method.ident.name == "is_empty" {
+                    span_lint_and_sugg(
+                        cx,
+                        NEEDLESS_COLLECT,
+                        expr.span,
+                        "you are collecting an iterator to check if it is empty",
+                        "consider replacing with",
+                        generate_needless_collect_is_empty_sugg(&args[0], cx),
+                    );
+                }
+                if method.ident.name == "contains" {
+                    span_lint_and_sugg(
+                        cx,
+                        NEEDLESS_COLLECT,
+                        expr.span,
+                        "you are collecting an iterator to check if contains an element",
+                        "consider replacing with",
+                        generate_needless_collect_contains_sugg(&args[0], &args[1], cx),
+                    );
+                }
+            }
+        }
+    }
+}
+
+fn generate_needless_collect_len_sugg<'a, 'tcx>(collect_expr: &'tcx Expr, cx: &LateContext<'a, 'tcx>) -> String {
+    if let ExprKind::MethodCall(_, _, ref args) = collect_expr.node {
+        let iter = snippet(cx, args[0].span, "??");
+        return format!("{}.count()", iter);
+    }
+    unreachable!();
+}
+
+fn generate_needless_collect_is_empty_sugg<'a, 'tcx>(collect_expr: &'tcx Expr, cx: &LateContext<'a, 'tcx>) -> String {
+    if let ExprKind::MethodCall(_, _, ref args) = collect_expr.node {
+        let iter = snippet(cx, args[0].span, "??");
+        return format!("{}.any(|_| true)", iter);
```

Comment:
Is that really the best suggestion? Feels a little hacky

Reply:
Hah, you're right. My mistake. I agree, `.next().is_none()` feels cleaner.